File: Scraper/elastic_client.js
var elasticsearch = require('elasticsearch');

var client = new elasticsearch.Client({
  host: 'localhost:9200'
});

module.exports.insert = function (data){
  has_link(data.link, function(link){
    if(JSON.stringify(link) != '{}'){
      console.log(JSON.stringify(link));

      if(has_change(data, link)){
        update(data, link);
      }
    }
    else{
      act_insert(data);
    }
  });
}

function act_insert(data){
  client.index({
  index: 'prices',
  type: 'products',
  body: data
},function(err,resp,status) {
    console.log(resp);
});
}

function has_link(l, callback){
  var ret_id = client.search({
    index: 'prices',
    type: 'products',
    body: {
      query: {
        bool : {
          must : {
            match : { link : l }
          }
        }
      }
    }
  }).then(function (resp) {
    var hits = resp.hits.hits;

    if(resp.hits.total > 0 ){
      var max_version = 0;
      var wanted_index = 0;
      var has_found = false;

      for(var i = 0; i < hits.length; i++){
          if(hits[i]._source.link == l && hits[i]._source.version > max_version){
            max_version = hits[i]._source.version;
            wanted_index = i;
            has_found = true;
          }
      }

      if(has_found){
        callback(hits[wanted_index]);
      }
      else{
        callback({});
      }
    }
    else{
      callback({});
    }
  }, function (err) {
    console.trace(err.message);
    callback({});
  });
}

function has_change(data, db_data){
  if(data.image == db_data._source.image &&
     data.price == db_data._source.price){
    return false;
  }
  return true;
}

function update(data, db_data){
  delete_from_id(db_data._id);
  var new_data = db_data._source;
  new_data.active = false;
  data.version++;

  act_insert(new_data);
  act_insert(data);
}

function delete_from_id(id){
  client.deleteByQuery({
        index: 'prices',
        type: 'products',
        body: {
          query: {
            bool : {
              must : {
                match : { _id : id }
              }
            }
          }
        }
    }, function (error, response) {
        console.log(JSON.stringify(error));
        console.log("Deleted");
    });
}

function get_from_id(id){
  client.search({
    index: 'prices',
    type: 'products',
    body: {
      query: {
        bool : {
          must : {
            match : { _id : id }
          },
          minimum_should_match : 1,
          boost : 1.0
        }
      }
    }
  }).then(function (resp) {
    var hits = resp.hits.hits;

    if(hits.length > 0 ){
      return hits[0];
    }
    else{
      return {};
    }
  }, function (err) {
    console.trace(err.message);
  });
}

File: PythonScraper/ElasticClient.py
import requests
import json
from elasticsearch5 import Elasticsearch

es = Elasticsearch([{'host': 'localhost', 'port':9200}])
res = requests.get('http://localhost:9200')

##býr til json document sem er sett í gagnasafnið
def CreateDocument(link, title, price, image, labels, category, features):
    doc = json.dumps({
        "title": title,
        "price": price,
        "image": image,
        "link": link,
        "labels": labels,
        "category": category,
        "features": features
        }, indent=4)
    return doc

##checkar hvort þetta sé til staðar í gagnasafninu ef svo þá breytir hann því þar ef þess þarf
def Insert(insertionData):
    jsonData = json.loads(insertionData)
    baseData = HasData(insertionData)

    if(baseData["hits"]["total"] > 0):
        inDb = HasChange(baseData, jsonData)

        if(inDb != {}):
            print("updated some data")
            Update(inDb, jsonData)
    else:
        print("inserted new data")
        ActualInsert(insertionData)

    return True

##setur data spurningarlaust inn í gagnasafnið
def ActualInsert(data):
    jsonData = json.loads(data)
    jsonData["version"] = 1
    jsonData["active"] = True

    res = es.index(index='prices', doc_type='products', body=json.dumps(jsonData))
    print(res)
    return True

##Í stað þess að setja inn í safnið þá frekar að uppfæra það ser er nú þegar til
def Update(baseData, data):
    newBaseData = baseData["_source"]
    newBaseData["active"] = False

    data["version"] = newBaseData["version"] + 1
    data["active"] = True
    
    es.delete(index='prices', doc_type='products', id=baseData["_id"])

    #es.index(index='prices', doc_type='products', body=newBaseData)
    es.index(index='prices', doc_type='products', body=data)
    
    return True

##checkar hvort það sé til í gagnasafninu
##ef svo þá returnar það datanu sem er til annars tómum streng
def HasData(data):
    jsonData = json.loads(data)
    
    searchDocument = json.dumps({
      "query": {
        "bool" : {
          "must" : {
            "match" : { "link" : jsonData["link"] }
          }
        }
      }
    }, indent=4)

    response = es.search(index='prices', body=searchDocument)

    return response

##checkar hvort það sé munur á hlutunum tvemur
def HasChange(data1 ,data2):

    for key in data1["hits"]["hits"]:
        print(key["_source"])

        if(key["_source"]["active"] == True):
            usedData = key
    
    if(usedData != {} and usedData["_source"]["price"] == data2["price"] and usedData["_source"]["title"] == data2["title"]):
        print("nei")
        return usedData

    return {}